Drift/Private/DriftBase.cpp

```diff
@@ -2995,44 +3009,56 @@ void FDriftBase::InitServerAuthentication()
         TEXT("{\"username\": \"%s\", \"password\": \"%s\", \"provider\": \"%s\"}"),
-        *SERVER_CREDENTIALS_USERNAME, *password, *SERVER_CREDENTIALS_PROVIDER
+        *SERVER_CREDENTIALS_USERNAME, *Password, *SERVER_CREDENTIALS_PROVIDER
         );
 
-    auto request = GetRootRequestManager()->Post(driftEndpoints.auth, payload, HttpStatusCodes::Ok);
+    const auto Request = GetRootRequestManager()->Post(driftEndpoints.auth, Payload, HttpStatusCodes::Ok);
 
-    DRIFT_LOG(Base, Verbose, TEXT("Authenticating server: %s"), *request->GetAsDebugString(true));
+    DRIFT_LOG(Base, Verbose, TEXT("Authenticating server: %s"), *Request->GetAsDebugString(true));
 
-    request->OnResponse.BindLambda([this](ResponseContext& context, JsonDocument& doc)
+    Request->OnResponse.BindLambda([this](ResponseContext& Context, JsonDocument& Doc)
     {
-        serverJTI_ = doc[TEXT("jti")].GetString();
-        serverBearerToken_ = doc[TEXT("token")].GetString();
+        serverJTI_ = Doc[TEXT("jti")].GetString();
+        serverBearerToken_ = Doc[TEXT("token")].GetString();
 
         if (serverBearerToken_.IsEmpty())
         {
-            context.error = TEXT("Session 'token' missing.");
+            Context.error = TEXT("Session 'token' missing.");
             return;
         }
         DRIFT_LOG(Base, Verbose, TEXT("Got bearer token %s"), *serverBearerToken_);
 
-        TSharedRef<JsonRequestManager> manager = MakeShareable(new JWTRequestManager(serverBearerToken_));
-        manager->DefaultErrorHandler.BindRaw(this, &FDriftBase::DefaultErrorHandler);
-        manager->DefaultDriftDeprecationMessageHandler.BindRaw(this, &FDriftBase::DriftDeprecationMessageHandler);
-        manager->SetApiKey(GetApiKeyHeader());
-        manager->SetCache(httpCache_);
-        SetGameRequestManager(manager);
-        eventManager->SetRequestManager(manager);
+        const TSharedRef<JsonRequestManager> Manager = MakeShareable(new JWTRequestManager(serverBearerToken_));
+        Manager->DefaultErrorHandler.BindRaw(this, &FDriftBase::DefaultErrorHandler);
+        Manager->DefaultDriftDeprecationMessageHandler.BindRaw(this, &FDriftBase::DriftDeprecationMessageHandler);
+        Manager->SetApiKey(GetApiKeyHeader());
+        Manager->SetCache(httpCache_);
+        SetGameRequestManager(Manager);
+        eventManager->SetRequestManager(Manager);
 
         InitServerRegistration();
     });
-    request->OnError.BindLambda([this](ResponseContext& context)
+    Request->OnError.BindLambda([this](ResponseContext& Context)
     {
-        // TODO: Error handling
-        context.errorHandled = true;
-        Reset();
+    	Context.errorHandled = true;
+
+		DRIFT_LOG(Base, Error, TEXT("Failed to authenticate server"));
+
+		if (Context.response.IsValid() && Context.response->GetResponseCode() >= 500)
+		{
+			DRIFT_LOG(Base, Log, TEXT("Retrying authenticating server"));
+
+			// Retry if server error (5XX HTTP status code)
+			InitServerAuthentication();
```

Comment:
Same here, don't just retry over and over, it can theoretically cause more load and more failures.